designs/http_client_design.md

```diff
@@ -0,0 +1,250 @@
+
+5. **Report outcome**
+   * Map `(resp, err)` into `request.Outcome` (success vs resource failure) and call `Tracker.ReportOutcome(ctx, clientID, outcome)`.
+
+**Retry interaction**
+
+* Retries are **optional**, configured via `RetryConfig`.
+* Default behavior:
+  * Retry only **idempotent methods** (`GET`, `HEAD`, `OPTIONS`, `TRACE`, `PUT`, `DELETE`).
+  * Retry on:
+    * Network errors.
+    * Configurable 5xx codes (default: 500, 502, 503, 504) and optionally 429.
+  * Use exponential backoff with jitter, with small, conservative defaults (e.g. max 3 attempts).
+
+> Important: **Each retry attempt** registers and reports its own outcome with FAIR, so client behavior remains visible to the fairness algorithm and can be throttled if necessary.
+
+**Standalone client**
+
+`FairHTTPClient`:
+
+* Constructs an `http.Transport` with production defaults:
+  * Higher `MaxIdleConns` / `MaxIdleConnsPerHost` than `stdlib` defaults.
+  * Sensible connection and header timeouts.
+* Wraps that transport with `FairRoundTripper`.
+* Exposes:
+  * `Do(req *http.Request) (*http.Response, error)`
+  * Convenience methods `Get`, `Post`, etc., delegating to the underlying `http.Client`.
+
+This lets users choose:
+
+* **Wrapper usage** (minimal change to existing code), or
+* **New client usage** (get FAIR + better transport defaults in one go).
+
+**Fail-Open vs Fail-Closed Behavior**
+
+If the client is unable to communicate with the FAIR tracker (e.g., local process issue, shared-memory corruption, or unexpected internal errors), we must choose between:
+
+#### Fail-Open (default)
+
+If FAIR errors, allow the request to proceed as a normal HTTP request, without fairness metadata.
+
+Rationale:
+- Prioritizes availability over fairness.
+- Prevents widespread outages if FAIR has a local transient issue.
+- Keeps FAIR as non-critical path for correctness.
+
+#### Fail-Closed (optional)
+
+If FAIR errors, return an error immediately and do not send the request.
+
+Rationale:
+- Prioritizes strict fairness guarantees.
+- Useful for environments where all traffic must be fairness-aware (internal-only systems, research testing, academic correctness verification).
+
+>
+> **Proposed Default: Fail-Open**
+> The majority of production use cases favor availability: if FAIR stops functioning momentarily, the system should degrade gracefully rather than blocking all outbound requests. Fail-closed remains available for advanced users who need stronger correctness guarantees.
+>
+
+### Security / Privacy
+
+**Client identifiers**
+
+* Document that **client IDs must not contain PII** (e.g. raw emails, password, tokens etc.)
+* Recommend using **opaque IDs or hashes**.
+* Provide `HashedClientIDExtractor` helper that wraps another extractor and returns a _SHA-256_ hash of the ID.
+
+FAIR already stores identifiers in a Bloom filter rather than as plain strings, but we still encourage non-PII identifiers for defense-in-depth.
```

Comment:
I think we are good here - so long as the headers are not logged we are fine here.